### PR TITLE
FourierTransform object, power spectrum, and average length scale analysis

### LINKED ISSUE #321: Add FFT power spectrum vector postprocessor
For studying patterning it would be nice to on the fly see the evolution of the power spectrum of the microstructure (and maybe its first moment). When using the guaranteed regular orthogonal MyTrimMesh this could be easily done by sampling the grid onto an FFTW3 array...

### PR DESCRIPTION
**Pull request by dschwen**
This still needs docs, but I want to see if testing works with FFTW3.

Closes #321

### CODE REVIEW

File: include/userobjects/FourierTransform.h [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+#include <memory>
+
+class FourierTransform;
+
+template <>
+InputParameters validParams<FourierTransform>();
+
+/**
+ * Compute the power spectrum of the fast fourier transform of a selected variable field.
```
Comment (snschune) [inline include/userobjects/FourierTransform.h]: forgot to update doc string here.

File: src/userobjects/FourierTransform.C [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+#include <functional>
+
+registerMooseObject("MagpieApp", FourierTransform);
+
+template <>
+InputParameters
+validParams<FourierTransform>()
+{
+  InputParameters params = validParams<ElementUserObject>();
+  params.addClassDescription("Compute the power spectrum of a given fast fourier transform. The "
```
Comment (snschune) [inline src/userobjects/FourierTransform.C]: This object computes the Fourier transform.

File: src/userobjects/FourierTransform.C [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+    // querying the mesh to set grid
+    for (unsigned int i = 0; i < _dim; ++i)
+      _grid.push_back(mytrim_mesh->getCellCountInDimension(i));
+  }
+
+  // refine grid by max_h_level powers of two
+  auto max_h_level = getParam<unsigned int>("max_h_level");
+  for (auto & count : _grid)
+    count = count << max_h_level;
```
Comment (snschune) [inline src/userobjects/FourierTransform.C]: does that do `count *= 2^max_h_level`?

Comment (dschwen) [reply src/userobjects/FourierTransform.C]: yes

**Pull request merged**